blockchain/chain/src/store/chain_store.rs

```diff
@@ -617,18 +594,63 @@ where
     //         None
     //     }
     // }
+}
 
-    /// Walks over tipset and state data and loads all blocks not yet seen.
-    /// This is tracked based on the callback function loading blocks.
-    async fn walk_snapshot<F>(
-        tipset: &Tipset,
-        recent_roots: ChainEpoch,
+/// Walks over tipset and state data and loads all blocks not yet seen.
+/// This is tracked based on the callback function loading blocks.
+struct SnapshotWalker<'db, 'chan, DB> {
+    skip_old_msgs: bool,
+    tx: &'chan flume::Sender<(Cid, Vec<u8>)>,
+    blockstore: &'db DB,
+}
+
+impl<'db, 'chan, DB> SnapshotWalker<'db, 'chan, DB>
+where
+    DB: Blockstore + Store + Send + Sync,
+{
+    fn new(
+        blockstore: &'db DB,
         skip_old_msgs: bool,
-        mut load_block: F,
-    ) -> Result<(), Error>
-    where
-        F: FnMut(Cid) -> Result<Vec<u8>, anyhow::Error>,
-    {
+        tx: &'chan flume::Sender<(Cid, Vec<u8>)>,
+    ) -> Self {
+        Self {
+            skip_old_msgs,
+            tx,
+            blockstore,
+        }
+    }
+
+    fn load(&self, cid: Cid) -> Result<Vec<u8>, Error> {
+        self.blockstore.get(&cid)?.ok_or_else(|| {
+            if self.skip_old_msgs {
+                Error::Other("Cid {cid} not found in blockstore".to_string())
+            } else {
+                Error::Other("Cid {cid} not found in blockstore. \
+                                Exporting a full snapshot is only possible when the node is initialised with a full one. \
+                                Consider exporting a lightweight snapshot, e.g. skip the old messages.".to_string())
+            }
+        })
+    }
+
+    async fn send_block(&self, cid: Cid) -> Result<(), Error> {
+        let block = self.load(cid)?;
+        self.tx
+            .send_async((cid, block))
+            .await
+            .expect("failed sending block for export");
+        Ok(())
+    }
```

Comment:
Avoiding cloning a block is not worth having duplicated code.

Reply:
Fixed.